core/src/main/scala/org/ensime/core/RichPresentationCompiler.scala

```diff
@@ -205,8 +206,84 @@ trait RichCompilerControl extends CompilerControl with RefactoringControl with C
   def askReloadAndTypeFiles(files: Iterable[SourceFile]) =
     askOption(reloadAndTypeFiles(files))
 
-  def askUsesOfSymAtPoint(p: Position): List[RangePosition] =
-    askOption(usesOfSymbolAtPoint(p).toList).getOrElse(List.empty)
+  def askUsesOfSymAtPos(pos: Position): List[RangePosition] = {
+    askLoadedTyped(pos.source)
+    val symbol = askSymbolAt(pos)
+    symbol match {
+      case None => Nil
+      case Some(sym) =>
+        val source = pos.source
+        val loadedFiles = loadUsesOfSym(sym)
+        val files = loadedFiles.map(_.file) + source.file.file.toPath
+        askUsesOfSym(sym, files)
+    }
+  }
+
+  def askUsesOfSym(sym: Symbol, files: collection.Set[Path]): List[RangePosition] =
+    askOption(usesOfSymbol(sym.pos, files).toList).getOrElse(List.empty)
+
+  protected def withExistingScalaFiles(
+    files: collection.Set[SourceFileInfo]
+  )(
+    f: List[SourceFile] => RpcResponse
+  ): RpcResponse = {
+    val (existing, missingFiles) = files.partition(_.exists())
+    if (missingFiles.nonEmpty) {
+      val missingFilePaths = missingFiles.map { f => "\"" + f.file + "\"" }.mkString(",")
+      EnsimeServerError(s"file(s): $missingFilePaths do not exist")
+    } else {
+      val scalas = existing.collect { case sfi @ SourceFileInfo(RawFile(path), _, _) if path.toString.endsWith(".scala") => sfi }
+      val sourceFiles: List[SourceFile] = scalas.map(createSourceFile)(collection.breakOut)
+      f(sourceFiles)
+    }
+  }
+
+  def handleReloadFiles(files: collection.Set[SourceFileInfo]): RpcResponse = {
+    withExistingScalaFiles(files) { scalas =>
+      if (scalas.nonEmpty) {
+        askReloadFiles(scalas)
+        askNotifyWhenReady()
+      }
+      VoidResponse
+    }
+  }
+
+  def handleReloadAndRetypeFiles(files: collection.Set[SourceFileInfo]): RpcResponse = {
+    withExistingScalaFiles(files) { scalas =>
+      if (scalas.nonEmpty) {
+        askReloadFiles(scalas)
+        scalas.foreach(askLoadedTyped)
+        askNotifyWhenReady()
+      }
+      VoidResponse
+    }
+  }
+
+  import org.ensime.util.file.File
+  def loadUsesOfSym(sym: Symbol): collection.Set[RawFile] = {
+    val files = usesOfSym(sym)
+    val sfis = files.map(rf => SourceFileInfo(rf))
+    handleReloadAndRetypeFiles(sfis)
+    files
+  }
+
+  def usesOfSym(sym: Symbol): collection.Set[RawFile] = {
+    import scala.concurrent.Await
+    import scala.concurrent.duration.Duration
+
+    val noReverseLookups = search.noReverseLookups
+    if (noReverseLookups) {
+      Set.empty
+    } else {
+      val fqn = askSymbolFqn(sym)
+      val usages = Await.result(search.findUsages(fqn.fqnString), Duration.Inf)
```

Comment:
yeah, this is where the blocking happens... I'm not too keen on this. It means the PC actor is blocking on DB access. It may be a big refactor though.